### ISSUE TITLE
Prevent busy-waiting in consensus gossip routines, when their channel's SendQueue is full

### ISSUE THREAD

**Issue by ValarDragon** [cometbft/cometbft#3414]
### Summary

The consensus gossip routines can busy-wait, when their outbound message queues are full. There are several cases where peer.Send fails, and we continue to the outer-loop, and we-will re-enter the queue to send more messages.

We observed what appears to be hitting this busy waiting in CPU profiles in Osmosis mainnet. It was observed when we had sustained 5MB blocks for ~15 minutes. We see tremendously high numbers of calls in gossipDataRoutine the proposal.ToProto, (and gossipDataRoutine in general) taking tremendous amounts of system resources.  (And same in Vote, but on a different hot-path for failing gossip)

Here is what we saw over one hour, with 15 minutes of 5MB blocks:

![image](https://github.com/cometbft/cometbft/assets/6440154/a6f2e355-9fb3-45a7-9cac-571e88d6fe07)

I have never visually noticed proposal.ToProto taking a long time in another prod benchmark before.

Here is what we typically see over one hour (at least relative proportions within here are on par with what I've seen over dozens of profiles):
![image](https://github.com/cometbft/cometbft/assets/6440154/08cf667c-03e6-40c4-a0c9-609a1996c3fd)

So the time (and codepath) looks like were busy waiting to me.

### Proposal

- [ ] Understand and fix however its possible that we filled our SendBuffer
  - Its of size 100, so the only thing I can think of its CPU times in outbound packet clearing (recent PR's I've made help improve) and related to @evanforbes findings on the OS not sending packets due to potential congestion control. (#3346 should help, but won't fix core issue)
- [ ] Prevent this edge case from ever busy waiting. (I'll make a PR)

**Comment by ValarDragon** [comment]
Turns out my BusyWaiting due to `Send` failure doesn't seem right, since I think Send blocks for 10s. I have no idea why we got this pattern then of so many .ToProto calls :thinking: 

I feel like this issue is some sort of BusyWaiting, we can repeatedly do this ToProto call

**Comment by cason** [comment]
> Turns out my BusyWaiting due to Send failure doesn't seem right,

Yes. When `Send` fails we already waited for a while and the peer is probably gone.

**Comment by cason** [comment]
I have in my backlog to check every `Send()` call in the consensus reactor and try to replace it by a `TrySend()` whenever possible.

**Comment by ValarDragon** [comment]
Ah this likely was busy-waiting (per https://github.com/cometbft/cometbft/pull/3412#issuecomment-2209377784 )
But busyWaiting due to `hasChannel` failing!

**Comment by cason** [comment]
Aha. But why would it fail in this case? Why is the node trying to send messages to a channel that either does not exist or the remote peer does not support?

**Comment by ValarDragon** [comment]
Not sure how we got to that point, but we should ensure that our peer has the data channel before starting the gossip routine to them then (or maybe we just refuse to add the peer to the reactor if they don't have all channels?)

This is definitely a dos attack vector that I guess we saw on mainnet

**Comment by cason** [comment]
I was checking here, but there is no way, from the `Peer` interface, to know whether a peer supports a given channel. I would add an interface method for that and use it to check, prior to running the sending routines at the reactors, if the peer indeed supports/listens to the corresponding channel.

**Comment by cason** [comment]
I think #3510 closes this issue, doesn't it?

**Comment by ValarDragon** [comment]
Yes it does, ty!
